table/tables/partition.go

```diff
@@ -387,157 +558,236 @@ func extractListPartitionExprColumns(ctx sessionctx.Context, pi *model.Partition
+}
+
+func (lp *ForListPruning) locateListColumnsPartitionByRow(ctx sessionctx.Context, r []types.Datum) (int, error) {
+	intersect := NewListPartitionLocationIntersect()
+	sc := ctx.GetSessionVars().StmtCtx
+	for _, colPrune := range lp.ColPrunes {
+		location, err := colPrune.LocatePartition(sc, r[colPrune.ExprCol.Index])
+		if err != nil {
+			return -1, errors.Trace(err)
+		}
+		if !intersect.Intersect(location) {
+			break
 		}
-		fmt.Fprintf(&buf, "(%s)", strings.Join(vs, ","))
 	}
-	buf.WriteString("))")
-	for _, vs := range def.InValues {
-		nullCondBuf.Reset()
-		hasNull := false
-		for i, value := range vs {
-			expr, err := parseSimpleExprWithNames(p, ctx, value, schema, names)
-			if err != nil {
-				return "", errors.Trace(err)
-			}
-			v, err := expr.Eval(chunk.Row{})
+	location := intersect.GetLocation()
+	if location.IsEmpty() {
+		return -1, table.ErrNoPartitionForGivenValue.GenWithStackByArgs("from column_list")
+	}
+	return location[0].PartIdx, nil
+}
+
+// buildListPartitionValueMap builds list columns partition value map for the specified column.
+// colIdx is the specified column index in the list columns.
+func (lp *ForListColumnPruning) buildPartitionValueMap(ctx sessionctx.Context, tblInfo *model.TableInfo, colIdx int,
+	schema *expression.Schema, names types.NameSlice, p *parser.Parser) error {
+	pi := tblInfo.GetPartitionInfo()
+	sc := ctx.GetSessionVars().StmtCtx
+	for partitionIdx, def := range pi.Definitions {
+		for groupIdx, vs := range def.InValues {
+			key, err := lp.genConstExprKey(ctx, sc, vs[colIdx], schema, names, p)
 			if err != nil {
-				return "", errors.Trace(err)
+				return errors.Trace(err)
 			}
-			if i > 0 {
-				nullCondBuf.WriteString(" and ")
-			}
-			if v.IsNull() {
-				hasNull = true
-				fmt.Fprintf(&nullCondBuf, "%s is null", pi.Columns[i])
-			} else {
-				fmt.Fprintf(&nullCondBuf, "%s = %s", pi.Columns[i], value)
+			location, ok := lp.valueMap[key]
+			if ok {
+				idx := location.findByPartitionIdx(partitionIdx)
+				if idx != -1 {
+					location[idx].GroupIdxs = append(location[idx].GroupIdxs, groupIdx)
+					continue
+				}
 			}
+			location = append(location, ListPartitionGroup{
+				PartIdx:   partitionIdx,
+				GroupIdxs: []int{groupIdx},
+			})
+			lp.valueMap[key] = location
 		}
-		if hasNull {
-			fmt.Fprintf(&buf, " or (%s) ", nullCondBuf.String())
-		}
 	}
-	return buf.String(), nil
+	return nil
+}
+
+func (lp *ForListColumnPruning) genConstExprKey(ctx sessionctx.Context, sc *stmtctx.StatementContext, exprStr string,
+	schema *expression.Schema, names types.NameSlice, p *parser.Parser) (string, error) {
+	expr, err := parseSimpleExprWithNames(p, ctx, exprStr, schema, names)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	v, err := expr.Eval(chunk.Row{})
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	key, err := lp.genKey(sc, v)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	return string(key), nil
```

Comment:
maybe return []byte like `genKey` is better :wink:

Reply:
great cache, done.